File: packages/gamepad/darwin/gamepad/Sources/gamepad/GamepadPlugin.swift
import GameController

#if os(macOS)
import FlutterMacOS
#else
import Flutter
import UIKit
#endif

/// Reads a controller through `GameController.framework` and forwards what it
/// says.
///
/// ## One source for macOS and iOS
///
/// The framework is the same framework on both, down to the property names, so
/// two copies of this file would be two copies to fix. `sharedDarwinSource` in
/// the pubspec is what lets one serve both; the only differences are which
/// Flutter module to import and which notification means "the player has gone
/// away", and both are three lines.
///
/// ## What is deliberately not here
///
/// **No arithmetic and no decisions**, the same rule the Android side follows.
/// The values go out in a fixed order and `lib/src/darwin_mapping.dart` decides
/// what they mean — including the one thing Apple does differently from
/// everybody else, which is that a stick's y is positive *upwards*. Flipping it
/// here would put it where no test could see it.
///
/// The order below is the order that file reads: six axes, then one value per
/// button in `PadButton.known`'s order. That list and `GCExtendedGamepad` agree
/// because both are the physical layout of a controller, and a test on the Dart
/// side checks the count.
public class GamepadPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private static let eventChannelName = "dev.flutter3d/gamepad/events"

  /// Six axes and seventeen buttons — see the class doc on the order.
  private static let axisCount = 6

  private var sink: FlutterEventSink?
  private weak var controller: GCController?
  private var observers: [NSObjectProtocol] = []

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = GamepadPlugin()
    #if os(macOS)
      let messenger = registrar.messenger
    #else
      let messenger = registrar.messenger()
    #endif
    let channel = FlutterEventChannel(name: eventChannelName, binaryMessenger: messenger)
    channel.setStreamHandler(instance)
  }

  // MARK: - Listening

  public func onListen(
    withArguments _: Any?, eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    sink = events
    observe()
    // Whatever is already paired. A controller connected before the game started
    // produces no notification and would otherwise stay invisible until the
    // player turned it off and on again.
    adopt(GCController.controllers().first)
    return nil
  }

  public func onCancel(withArguments _: Any?) -> FlutterError? {
    for observer in observers {
      NotificationCenter.default.removeObserver(observer)
    }
    observers.removeAll()
    controller?.extendedGamepad?.valueChangedHandler = nil
    controller = nil
    sink = nil
    return nil
  }

  private func observe() {
    let centre = NotificationCenter.default
    observers.append(
      centre.addObserver(
        forName: .GCControllerDidConnect, object: nil, queue: .main
      ) { [weak self] note in
        self?.adopt(note.object as? GCController)
      })

    observers.append(
      centre.addObserver(
        forName: .GCControllerDidDisconnect, object: nil, queue: .main
      ) { [weak self] note in
        guard let self, (note.object as? GCController) === self.controller else { return }
        self.controller = nil
        // The Dart side zeroes before it passes the news on, so a controller
        // whose battery dies mid-corner cannot leave the throttle where it was.
        self.sink?(["event": "disconnected"])
        // Another pad may still be paired. Taking it is what makes swapping
        // controllers work without a relaunch.
        self.adopt(GCController.controllers().first)
      })

    // Losing the player, which is not the same as losing the pad. Handled
    // natively, as `MouseCapturePlugin` handles focus loss and for its reason:
    // the last values before the window went away are the ones the pad would
    // otherwise stay in, and a stick left half over keeps walking behind
    // whatever is now on screen.
    #if os(macOS)
      let away = NSApplication.didResignActiveNotification
    #else
      let away = UIApplication.willResignActiveNotification
    #endif
    observers.append(
      centre.addObserver(forName: away, object: nil, queue: .main) { [weak self] _ in
        self?.sink?(["event": "relaxed"])
      })
  }

  // MARK: - The controller

  private func adopt(_ candidate: GCController?) {
    guard let candidate, let pad = candidate.extendedGamepad else { return }
    guard candidate !== controller else { return }

    controller?.extendedGamepad?.valueChangedHandler = nil
    controller = candidate
    sink?([
      "event": "connected",
      "name": candidate.vendorName ?? "",
    ])

    pad.valueChangedHandler = { [weak self] pad, _ in
      self?.forward(pad)
    }
    // Once immediately: the handler fires on change, and a stick already held
    // when the game started has not changed since.
    forward(pad)
  }

  private func forward(_ pad: GCExtendedGamepad) {
    guard let sink else { return }

    // Fixed order, and the Dart side reads it by index. Sticks first, then the
    // triggers as travel, then a value per button.
    var sample: [Double] = [
      Double(pad.leftThumbstick.xAxis.value),
      Double(pad.leftThumbstick.yAxis.value),
      Double(pad.rightThumbstick.xAxis.value),
      Double(pad.rightThumbstick.yAxis.value),
      Double(pad.leftTrigger.value),
      Double(pad.rightTrigger.value),
    ]
    sample.reserveCapacity(GamepadPlugin.axisCount + 17)

    // `PadButton.known`'s order, which is the physical layout: the four face
    // buttons by position, the d-pad, the shoulders, the triggers, the stick
    // clicks, then menu, options and home.
    let buttons: [GCControllerButtonInput?] = [
      pad.buttonA, pad.buttonB, pad.buttonX, pad.buttonY,
      pad.dpad.up, pad.dpad.down, pad.dpad.left, pad.dpad.right,
      pad.leftShoulder, pad.rightShoulder,
      pad.leftTrigger, pad.rightTrigger,
      pad.leftThumbstickButton, pad.rightThumbstickButton,
      pad.buttonMenu, pad.buttonOptions, pad.buttonHome,
    ]
    for button in buttons {
      guard let button else {
        // A control this pad does not have — plenty have no home button and
        // older ones have no clickable sticks. Nought rather than a gap, so the
        // indices stay where the Dart side expects them.
        sample.append(0.0)
        continue
      }
      // A bit, and only a bit — the framework's own `isPressed`, so nobody here
      // decides what counts as pressed. A trigger's travel is already in the
      // axes above and does not need sending twice.
      sample.append(button.isPressed ? 1.0 : 0.0)
    }

    sink(FlutterStandardTypedData(float64: Data(bytes: sample, count: sample.count * 8)))
  }
}

File: packages/pad_input/android/src/main/kotlin/dev/flutter3d/gamepad/GamepadPlugin.kt
package dev.flutter3d.gamepad

import android.app.Activity
import android.app.Application
import android.hardware.input.InputManager
import android.os.Bundle
import android.view.InputDevice
import android.view.MotionEvent
import android.view.View

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel

/**
 * Forwards a gamepad's stick and trigger motion, and says which pad is attached.
 *
 * ## What is deliberately not here
 *
 * **No arithmetic and no decisions.** Which axis a trigger is on, whether the
 * d-pad is a hat or a pair of keys, what "pressed" means for an analogue
 * control: all of it is in `lib/src/android_mapping.dart`, in Dart, where a unit
 * test can reach it. This file reports the axes a device admits to having and the
 * values it sends, and nothing else.
 *
 * That split is the reason an Android backend could be written here at all. The
 * specification records that an earlier gamepad backend was deleted rather than
 * finished, because one written without a controller in hand is wrong in a way no
 * test shows — and the way to survive that rule is to leave native code with
 * nothing in it that a test would have caught.
 *
 * **No buttons either.** Android delivers a gamepad's buttons as `KeyEvent`s, and
 * Flutter's own embedding already maps their key codes to `gameButtonA` and its
 * neighbours and hands them to the framework. Catching them again here would be a
 * second source of truth for one event, and the two would disagree the first time
 * either was wrong.
 *
 * ## Why a listener on the decor view
 *
 * Flutter turns pointer and scroll motion into `PointerEvent`s and drops the
 * rest, so a joystick's `MotionEvent` never reaches Dart. A plugin cannot
 * override `Activity.onGenericMotionEvent`, but `View.setOnGenericMotionListener`
 * is consulted before a view handles an event itself, and the window's decor view
 * sees what the activity sees. That is the whole hook.
 */
class GamepadPlugin :
    FlutterPlugin,
    ActivityAware,
    EventChannel.StreamHandler,
    InputManager.InputDeviceListener {

  private companion object {
    const val EVENT_CHANNEL = "dev.flutter3d/gamepad/events"

    /**
     * The axes Dart asks about, in the order it will read them.
     *
     * Kept in sync with `AndroidAxis.all` by hand, and that is on purpose: the
     * alternative is a channel call at startup to fetch a list, which is a round
     * trip and a failure mode in exchange for a list that changes when a mapping
     * changes and not otherwise.
     */
    val AXES = intArrayOf(
      MotionEvent.AXIS_X,
      MotionEvent.AXIS_Y,
      MotionEvent.AXIS_Z,
      MotionEvent.AXIS_RZ,
      MotionEvent.AXIS_HAT_X,
      MotionEvent.AXIS_HAT_Y,
      MotionEvent.AXIS_LTRIGGER,
      MotionEvent.AXIS_RTRIGGER,
      MotionEvent.AXIS_GAS,
      MotionEvent.AXIS_BRAKE,
    )
  }

  private var eventChannel: EventChannel? = null
  private var sink: EventChannel.EventSink? = null
  private var activity: Activity? = null
  private var inputManager: InputManager? = null

  /** The pad being reported, or null. One at a time, as the package promises. */
  private var deviceId: Int? = null

  /** The axes [deviceId] said it has, so a sample carries only those. */
  private var deviceAxes: IntArray = IntArray(0)

  private val sample = DoubleArray(1 + AXES.size * 2)

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL).also {
      it.setStreamHandler(this)
    }
    inputManager =
      binding.applicationContext.getSystemService(Application.INPUT_SERVICE)
          as? InputManager
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    // The same teardown onCancel does, because the engine can go while the
    // stream is still listening — a hot restart never cancels it — and this
    // object would otherwise stay registered with the InputManager and on the
    // decor view for as long as the process lives. Both calls are no-ops when
    // onCancel already ran.
    inputManager?.unregisterInputDeviceListener(this)
    detachMotionListener()
    sink = null
    eventChannel?.setStreamHandler(null)
    eventChannel = null
    inputManager = null
  }

  // MARK: - Listening

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    sink = events
    inputManager?.registerInputDeviceListener(this, null)
    attachMotionListener()
    // Whatever is already plugged in, because a controller connected before the
    // game started produces no event and would otherwise be invisible until the
    // player unplugged and replugged it.
    adopt(findGamepad())
  }

  override fun onCancel(arguments: Any?) {
    inputManager?.unregisterInputDeviceListener(this)
    detachMotionListener()
    sink = null
  }

  // MARK: - Devices

  override fun onInputDeviceAdded(id: Int) {
    if (isGamepad(InputDevice.getDevice(id))) adopt(id)
  }

  override fun onInputDeviceRemoved(id: Int) {
    if (id != deviceId) return
    deviceId = null
    deviceAxes = IntArray(0)
    // Told after the fact, and the Dart side zeroes before it passes the news on:
    // a pad whose battery dies mid-corner must not leave the throttle where it
    // was.
    sink?.success(mapOf("event" to "disconnected"))
  }

  override fun onInputDeviceChanged(id: Int) {
    // A device that gained or lost axes is a different device as far as the
    // mapping is concerned — a keyboard with a joystick dock, in practice.
    if (id == deviceId) adopt(id)
  }

  /** The first attached controller, or null. */
  private fun findGamepad(): Int? =
    InputDevice.getDeviceIds().firstOrNull { isGamepad(InputDevice.getDevice(it)) }

  /**
   * Whether this is something a player holds.
   *
   * Both source bits, because they disagree: a pad with only a d-pad and buttons
   * reports `SOURCE_GAMEPAD` without `SOURCE_JOYSTICK`, and a flight stick the
   * other way round. Requiring both would refuse half the hardware.
   */
  private fun isGamepad(device: InputDevice?): Boolean {
    if (device == null || device.isVirtual) return false
    val sources = device.sources
    val gamepad = sources and InputDevice.SOURCE_GAMEPAD == InputDevice.SOURCE_GAMEPAD
    val joystick = sources and InputDevice.SOURCE_JOYSTICK == InputDevice.SOURCE_JOYSTICK
    return gamepad || joystick
  }

  private fun adopt(id: Int?) {
    if (id == null) return
    val device = InputDevice.getDevice(id) ?: return
    // What it actually has, asked rather than assumed. This is the answer Dart
    // needs to know whether a trigger is on `AXIS_LTRIGGER` or on `AXIS_BRAKE`,
    // and the two are the same physical control under different drivers.
    deviceAxes = AXES.filter { device.getMotionRange(it, device.sources) != null }
      .toIntArray()
    deviceId = id
    sink?.success(
      mapOf(
        "event" to "connected",
        "device" to id,
        "name" to (device.name ?: ""),
        "axes" to deviceAxes.toList(),
      )
    )
  }

  // MARK: - Motion

  private val motionListener = View.OnGenericMotionListener { _, event ->
    forward(event)
    // **False, always.** Claiming the event would stop it reaching anything else
    // that wanted it, and this plugin is a listener rather than an owner. It also
    // costs nothing: nothing else in a Flutter application reads joystick motion.
    false
  }

  private fun attachMotionListener() {
    activity?.window?.decorView?.setOnGenericMotionListener(motionListener)
  }

  private fun detachMotionListener() {
    activity?.window?.decorView?.setOnGenericMotionListener(null)
  }

  private fun forward(event: MotionEvent) {
    val sink = this.sink ?: return
    if (event.deviceId != deviceId) return
    if (event.source and InputDevice.SOURCE_CLASS_JOYSTICK == 0) return

    sample[0] = event.deviceId.toDouble()
    var at = 1
    for (axis in deviceAxes) {
      sample[at++] = axis.toDouble()
      sample[at++] = event.getAxisValue(axis).toDouble()
    }
    // One typed list rather than a map, because this is the hot path: a stick in
    // motion sends a sample per frame and a `HashMap` per frame with it would be
    // the only allocation in the whole package that scales with play. The Dart
    // side tells the messages apart by type, as `pointer_lock` does.
    sink.success(sample.copyOfRange(0, at))
  }

  // MARK: - Activity, and losing it

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    if (sink != null) attachMotionListener()
    binding.activity.application
      .registerActivityLifecycleCallbacks(lifecycle)
  }

  override fun onDetachedFromActivity() {
    detachMotionListener()
    activity?.application?.unregisterActivityLifecycleCallbacks(lifecycle)
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
    onAttachedToActivity(binding)

  override fun onDetachedFromActivityForConfigChanges() = onDetachedFromActivity()

  /**
   * Releases everything when the application goes to the background.
   *
   * Handled natively, as `pointer_lock` handles focus loss, and for its reason:
   * the last event before the window went away is the state the pad stays in
   * otherwise, so a stick left half over keeps walking behind whatever is now on
   * screen. The pad is reported as still attached, because it is — the player is
   * what has gone.
   */
  private val lifecycle = object : Application.ActivityLifecycleCallbacks {
    override fun onActivityPaused(paused: Activity) {
      if (paused == activity) sink?.success(mapOf("event" to "relaxed"))
    }

    override fun onActivityCreated(a: Activity, state: Bundle?) {}
    override fun onActivityStarted(a: Activity) {}
    override fun onActivityResumed(a: Activity) {}
    override fun onActivityStopped(a: Activity) {}
    override fun onActivitySaveInstanceState(a: Activity, state: Bundle) {}
    override fun onActivityDestroyed(a: Activity) {}
  }
}
